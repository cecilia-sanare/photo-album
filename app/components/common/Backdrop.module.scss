.backdrop {
    background: rgba(0, 0, 0, 0.25);
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    opacity: 0;
    z-index: 1;
    
    transition: 0.15s ease-in-out;
    transition-property: opacity;
    
    &:not(.opened) {
        pointer-events: none;
    }

    &.opened {
        opacity: 1;
    }
}