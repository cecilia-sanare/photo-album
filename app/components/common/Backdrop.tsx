import classNames from 'classnames';
import * as styles from './Backdrop.module.scss';
import { useCachedState } from '../../hooks/use-cached-state';

type BackdropProps = {
  opened?: boolean;
  onClose?: () => void;
}

export function Backdrop({
  opened = false,
  onClose
}: BackdropProps) {
  const [internallyOpened, setInternallyOpened] = useCachedState<boolean>(() => opened, [opened]);

  return (
    <div
      data-testid='backdrop'
      className={classNames(
        styles.backdrop,
        internallyOpened && styles.opened
      )}
      onClick={() => {
        setInternallyOpened(false);
        onClose?.();
      }}
    />
  )
}
