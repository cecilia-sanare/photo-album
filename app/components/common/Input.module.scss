.input {
    border-radius: 3px;
    border: none;
    padding: 0 10px;
    min-height: 40px;
    outline: none;
    color: black;
    background: rgba(255, 255, 255, 0.8);

    transition: 0.15s ease-in-out;
    transition-property: background-color;

    &::placeholder {
        color: rgba(0, 0, 0, 0.5);
    }

    &:hover,
    &:active,
    &:focus {
        background: rgba(255, 255, 255, 0.9);
    }
}