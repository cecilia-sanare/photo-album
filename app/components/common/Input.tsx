import React from 'react';
import * as styles from './Input.module.scss';
import classNames from 'classnames';
import { useCachedState } from '../../hooks/use-cached-state';

type InputProps = {
  className?: string;
  value?: string;
  onChange?: (value: string) => void;
} & React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>;

export function Input({
  className,
  value,
  onChange,
  ...extraProps
}: InputProps) {
  const [internalValue, setInternalValue] = useCachedState<string | null>(() => value, [value]);

  return (
    <input
      className={classNames(
        styles.input,
        className
      )}
      value={internalValue}
      onChange={(event) => {
        setInternalValue(event.target.value);
        onChange?.(event.target.value);
      }}
      {...extraProps}
    />
  );
}
