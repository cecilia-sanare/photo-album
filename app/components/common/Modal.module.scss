.modal {
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    margin: 20px;
    padding: 10px;
    background: #2f3640;
    opacity: 0;
    z-index: 1;
    gap: 10px;
    display: flex;
    flex-direction: column;
    border-radius: 3px;
    max-width: 1000px;
    margin: 20px auto;
    box-sizing: border-box;
    
    transition: 0.15s ease-in-out;
    transition-property: opacity;

    &.opened {
        opacity: 1;
    }

    &:not(.opened) {
        pointer-events: none;
    }
}