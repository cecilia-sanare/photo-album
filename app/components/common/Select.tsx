import * as styles from './Select.module.scss';
import { useCachedState } from '../../hooks/use-cached-state';

type Primitive = string | number | boolean

export type SelectItem<T extends Primitive> = {
  label: string;
  value: T;
}

type SelectProps<T extends Primitive> = {
  items: SelectItem<T>[];
  value?: T;
  placeholder?: string;
  onChange?: (value: T) => void;
}

export function Select<T extends Primitive>({
  items,
  value,
  placeholder,
  onChange
}: SelectProps<T>) {
  const [selectedValue, setSelectedValue] = useCachedState<string>(() => value?.toString() ?? '', [value]);

  return (
    <select
      className={styles.select}
      data-testid='select'
      value={selectedValue}
      onChange={(event) => {
        const match = items.find((item) => item.value.toString() === event.target.value);

        setSelectedValue(match.value.toString());
        onChange?.(match.value);
      }}
    >
      <option value='' disabled>{placeholder}</option>
      {items.map((item) => (
        <option key={item.value.toString()} value={item.value.toString()}>{item.label}</option>
      ))}
    </select>
  )
}
