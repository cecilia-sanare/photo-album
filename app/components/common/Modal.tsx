import classNames from 'classnames';
import * as styles from './Modal.module.scss';
import { ReactNode, useEffect } from 'react';
import { Backdrop } from './Backdrop';
import { useCachedState } from '../../hooks/use-cached-state';

type ModalProps = {
  opened?: boolean;
  children: ReactNode;
  onClose?: () => void;
}

export function Modal({
  opened = false,
  children,
  onClose
}: ModalProps) {
  const [internallyOpened, setInternallyOpened] = useCachedState<boolean>(() => opened, [opened]);

  useEffect(() => {
    if (internallyOpened) {
      document.body.classList.add('no-scroll');
    } else {
      document.body.classList.remove('no-scroll');
    }
  }, [internallyOpened]);

  return (
    <>
      <Backdrop
        opened={internallyOpened}
        onClose={() => {
          setInternallyOpened(false);
          onClose?.();
        }}
      />
      <div
        className={classNames(
          styles.modal,
          internallyOpened && styles.opened
        )}
        data-testid='modal'
      >
        {children}
      </div>
    </>
  )
}
