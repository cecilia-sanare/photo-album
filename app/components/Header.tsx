import React, { useEffect } from 'react';
import { useAppDispatch } from '../store/store';
import { Input } from './common/Input';
import { Select } from './common/Select';
import * as styles from './Header.module.scss';
import { fetchAlbums, selectAlbums } from '../store/slices/albums.slice';
import { useSelector } from 'react-redux';
import { selectSearchQuery, selectSelectedAlbumId, setSearchQuery, setSelectedAlbumId } from '../store/slices/search.slice';
import { useReadOnlyCachedState } from '../hooks/use-cached-state';

export function Header() {
  const dispatch = useAppDispatch();
  const albums = useSelector(selectAlbums);
  const searchQuery = useSelector(selectSearchQuery);
  const selectedAlbumId = useSelector(selectSelectedAlbumId);
  const items = useReadOnlyCachedState(() => {
    return albums.map((album) => ({
      label: album.title,
      value: album.id
    }));
  }, [albums]);

  useEffect(() => {
    dispatch(fetchAlbums());
  }, []);

  return (
    <div className={styles.header}>
      <div className={styles.content}>
        <Input
          className={styles.search}
          placeholder='Query the photo names~'
          value={searchQuery}
          onChange={(query) => dispatch(setSearchQuery(query))}
        />
        <Select
          items={items}
          placeholder='Select an Album'
          value={selectedAlbumId}
          onChange={(albumId) => dispatch(setSelectedAlbumId(albumId))}
        />
      </div>
    </div>
  );
}
