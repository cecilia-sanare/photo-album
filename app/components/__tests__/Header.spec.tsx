import { fireEvent, render } from '@testing-library/react';
import { Header } from '../Header';
import { Provider } from 'react-redux';
import { store } from '../../store/store';
import { Chance } from 'chance';
import userEvent from '@testing-library/user-event';
import { SelectItem } from '../common/Select';
import axios from 'axios';

jest.mock('axios');
const axiosMocked = jest.mocked(axios);

const chance = new Chance();

describe('<Header />', () => {
  beforeEach(() => {
    jest.spyOn(store, 'dispatch').mockImplementation(() => {});
  });

  afterEach(() => {
    jest.resetAllMocks();
  })

  it('should render the Header', () => {
    const component = render(
      <Provider store={store}>
        <Header />
      </Provider>
    );

    expect(component.baseElement).toMatchSnapshot();
  });

  it('should request the albums', () => {
    render(
      <Provider store={store}>
        <Header />
      </Provider>
    );

    expect(store.dispatch).toHaveBeenCalledTimes(1);
  });

  describe('when a search query is entered', () => {
    it('should update the search query in state', async () => {
      const component = render(
        <Provider store={store}>
          <Header />
        </Provider>
      );

      await fireEvent.change(component.getByTestId('input'), {target: {value: chance.string()}})

      expect(store.dispatch).toHaveBeenCalledTimes(2);
    });
  });

  describe('when an album is selected', () => {
    it('should update the selected album', async () => {
      store.getState().albums = [{
        id: chance.integer(),
        userId: chance.integer(),
        title: chance.word()
      }];

      const component = render(
        <Provider store={store}>
          <Header />
        </Provider>
      );

      await userEvent.selectOptions(component.getByTestId('select'), store.getState().albums[0].id.toString());

      expect(store.dispatch).toHaveBeenCalledTimes(2);
    });
  });
});
