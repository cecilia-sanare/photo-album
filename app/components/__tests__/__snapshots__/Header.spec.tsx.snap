// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should render the Header 1`] = `
<body>
  <div>
    <div
      class="header"
    >
      <div
        class="content"
      >
        <input
          class="input search"
          data-testid="input"
          placeholder="Query the photo names~"
          value=""
        />
        <select
          class="select"
          data-testid="select"
        >
          <option
            disabled=""
            value=""
          >
            Select an Album
          </option>
        </select>
      </div>
    </div>
  </div>
</body>
`;
